File: carapace/receipt_protocol.py
"""Receipt v2 wire contract. Mirrored in ARIA; shared vectors guard parity."""
from __future__ import annotations

import hashlib
import math
import re
from datetime import datetime
from typing import Any

import jcs

DOMAIN = "carapace.tool-receipt.v2"
FIELDS = (
    "receipt_version", "domain", "receipt_id", "issued_at", "tool_id",
    "tool_version", "agent_id", "authorization_id", "args_hash", "result_hash",
    "status", "public_key",
)
HEX64 = re.compile(r"[0-9a-f]{64}\Z")
HEX128 = re.compile(r"[0-9a-f]{128}\Z")


def _json_value(value: Any) -> None:
    if value is None or isinstance(value, bool):
        return
    if isinstance(value, str):
        value.encode("utf-8", errors="strict")
        return
    if isinstance(value, (int, float)):
        if not math.isfinite(value) or (value == int(value) and abs(value) > 2**53 - 1):
            raise ValueError("Receipt JSON numbers must be finite and integers must be safe")
        return
    if isinstance(value, list):
        for item in value:
            _json_value(item)
        return
    if isinstance(value, dict) and all(isinstance(key, str) for key in value):
        for key, item in value.items():
            _json_value(key)
            _json_value(item)
        return
    raise ValueError("Receipt content must be JSON; implicit string conversion is forbidden")


def canonical_bytes(value: Any) -> bytes:
    _json_value(value)
    return jcs.canonicalize(value)


def content_hash(value: Any) -> str:
    return hashlib.sha256(canonical_bytes(value)).hexdigest()


def receipt_payload(receipt: dict) -> dict:
    if set(receipt) != set(FIELDS) | {"call_hash", "signature"}:
        raise ValueError("Incomplete or unknown receipt fields")
    if receipt["receipt_version"] != "2" or receipt["domain"] != DOMAIN:
        raise ValueError("Unsupported receipt version/domain")
    for field in ("receipt_id", "issued_at", "tool_id"):
        if not isinstance(receipt[field], str) or not receipt[field] or len(receipt[field]) > 512:
            raise ValueError(f"Invalid {field}")
    for field in ("agent_id", "tool_version", "authorization_id"):
        value = receipt[field]
        if value is not None and (not isinstance(value, str) or not value or len(value) > 512):
            raise ValueError(f"Invalid {field}")
    if not re.fullmatch(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})", receipt["issued_at"]):
        raise ValueError("Receipt time must be an ISO timestamp with timezone")
    issued = datetime.fromisoformat(receipt["issued_at"].replace("Z", "+00:00"))
    if issued.tzinfo is None:
        raise ValueError("Receipt time must include timezone")
    if receipt["status"] not in ("ok", "error"):
        raise ValueError("Invalid receipt status")
    for field in ("args_hash", "call_hash"):
        if not isinstance(receipt[field], str) or not HEX64.fullmatch(receipt[field]):
            raise ValueError(f"Invalid {field}")
    for field in ("result_hash", "public_key"):
        if receipt[field] is not None and (not isinstance(receipt[field], str) or not HEX64.fullmatch(receipt[field])):
            raise ValueError(f"Invalid {field}")
    sig = receipt["signature"]
    if sig is not None and (not isinstance(sig, str) or not HEX128.fullmatch(sig)):
        raise ValueError("Invalid signature")
    if (sig is None) != (receipt["public_key"] is None):
        raise ValueError("Signature and public key must occur together")
    return {field: receipt[field] for field in FIELDS}


def validate_receipt(receipt: dict, public_key_hex: str | None = None) -> bool:
    """Validate v2 integrity. True for unsigned v2; this is NOT issuer trust."""
    try:
        payload = receipt_payload(receipt)
        if content_hash(payload) != receipt["call_hash"]:
            return False
        if public_key_hex is not None and public_key_hex != receipt["public_key"]:
            return False
        if receipt["signature"] is not None:
            from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
            key = Ed25519PublicKey.from_public_bytes(bytes.fromhex(receipt["public_key"]))
            key.verify(bytes.fromhex(receipt["signature"]), receipt["call_hash"].encode("ascii"))
        return True
    except (ValueError, TypeError, KeyError, OverflowError, UnicodeError):
        return False
    except Exception:
        # Cryptography's InvalidSignature is deliberately returned as a negative result.
        return False

File: typescript/src/receipt-protocol.ts
import canonicalize from 'canonicalize';

export const DOMAIN = 'carapace.tool-receipt.v2';
export const FIELDS = [
  'receipt_version', 'domain', 'receipt_id', 'issued_at', 'tool_id',
  'tool_version', 'agent_id', 'authorization_id', 'args_hash', 'result_hash',
  'status', 'public_key',
] as const;

function jsonValue(value: unknown): void {
  if (value === null || typeof value === 'boolean') return;
  if (typeof value === 'string') {
    // Reject isolated UTF-16 surrogates; UTF-8 replacement would change the data.
    for (const character of value) {
      const point = character.codePointAt(0)!;
      if (point >= 0xd800 && point <= 0xdfff) throw new Error('Invalid Unicode');
    }
    return;
  }
  if (typeof value === 'number') {
    if (!Number.isFinite(value) || (Number.isInteger(value) && !Number.isSafeInteger(value))) {
      throw new Error('Receipt JSON numbers must be finite and integers must be safe');
    }
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) jsonValue(item);
    return;
  }
  if (typeof value === 'object' && value !== null &&
      (Object.getPrototypeOf(value) === Object.prototype || Object.getPrototypeOf(value) === null)) {
    for (const [key, item] of Object.entries(value)) { jsonValue(key); jsonValue(item); }
    return;
  }
  throw new Error('Receipt content must be JSON');
}

export async function contentHash(value: unknown): Promise<string> {
  jsonValue(value);
  const text = canonicalize(value);
  if (text === undefined) throw new Error('Invalid JSON');
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(text));
  return Array.from(new Uint8Array(digest), b => b.toString(16).padStart(2, '0')).join('');
}

export function timestampMillis(value: unknown): number {
  if (typeof value !== 'string') throw new Error('Invalid timestamp');
  const match = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})$/.exec(value);
  if (!match) throw new Error('Timestamp requires timezone');
  const [, year, month, day, hour, minute, second] = match.map(Number);
  const leap = year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
  const days = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  const result = Date.parse(value);
  if (year < 1 || month < 1 || month > 12 || day < 1 || day > days[month - 1] ||
      hour > 23 || minute > 59 || second > 59 || !Number.isFinite(result)) throw new Error('Invalid timestamp');
  return result;
}

export function receiptPayload(receipt: Record<string, unknown>): Record<string, unknown> {
  const expected = [...FIELDS, 'call_hash', 'signature'];
  if (Object.keys(receipt).length !== expected.length || expected.some(k => !(k in receipt))) {
    throw new Error('Incomplete or unknown receipt fields');
  }
  if (receipt.receipt_version !== '2' || receipt.domain !== DOMAIN) throw new Error('Unsupported receipt');
  for (const field of ['receipt_id', 'issued_at', 'tool_id']) {
    const value = receipt[field];
    if (typeof value !== 'string' || !value || Array.from(value).length > 512) throw new Error(`Invalid ${field}`);
  }
  for (const field of ['agent_id', 'tool_version', 'authorization_id']) {
    const value = receipt[field];
    if (value !== null && (typeof value !== 'string' || !value || Array.from(value).length > 512)) throw new Error(`Invalid ${field}`);
  }
  timestampMillis(receipt.issued_at);
  if (receipt.status !== 'ok' && receipt.status !== 'error') throw new Error('Invalid status');
  for (const field of ['args_hash', 'call_hash', 'result_hash', 'public_key']) {
    const value = receipt[field];
    if (value === null && (field === 'result_hash' || field === 'public_key')) continue;
    if (typeof value !== 'string' || !/^[0-9a-f]{64}$/.test(value)) throw new Error(`Invalid ${field}`);
  }
  if (receipt.signature !== null && (typeof receipt.signature !== 'string' || !/^[0-9a-f]{128}$/.test(receipt.signature))) throw new Error('Invalid signature');
  if ((receipt.signature === null) !== (receipt.public_key === null)) throw new Error('Incomplete signature');
  return Object.fromEntries(FIELDS.map(field => [field, receipt[field]]));
}
